src/lib.rs

```diff
@@ -22,23 +22,42 @@ However, CGS has a number of advantages:
 
 use std::marker::PhantomData;
 use std::{mem, ops};
-use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard, RwLockWriteGuard, Weak};
 
 /// Reference counter type. It doesn't make sense to allocate too much bit for it in regular applications.
 // TODO: control by a cargo feature
 type RefCount = u16;
 
+/// Epoch type determines the number of overwrites of components in storage.
+/// TODO: control by a cargo feature
+type Epoch = u16;
+
+/// The error type which is returned from upgrading WeakPointer.
+pub enum UpgradeErr {
+    /// Storage has been dropped.
+    DeadStorage,
+    /// Specific component in storage has been dropped or there are no `Pointer`s to it.
+    DeadComponent,
+}
+
 /// Inner storage data that is locked by `RwLock`.
 struct StorageInner<T> {
     data: Vec<T>,
     meta: Vec<RefCount>,
-    free_list: Vec<usize>,
+    free_list: Vec<(usize, Epoch)>,
 }
 
 /// Pending reference counts updates.
 struct Pending {
     add_ref: Vec<usize>,
     sub_ref: Vec<usize>,
+    epoch: Vec<Epoch>,
+}
+
+impl Pending {
+    fn prepare_sub_drain(&mut self) -> (&mut Vec<usize>, &Vec<Epoch>) {
```

Comment:
let's return `(Drain<usize>, &[Epoch])` pair here and rename the method to `drain_sub`